modin/distributed/dataframe/pandas/partitions.py

```diff
@@ -158,9 +175,20 @@ def from_partitions(partitions, axis):
         raise ValueError(
             f"Got unacceptable value of axis {axis}. Possible values are {0}, {1} or {None}."
         )
+    if index is None:
+        index = partition_mgr_class.get_indices(0, parts, lambda df: df.axes[0])
+
+    if columns is None:
+        columns = partition_mgr_class.get_indices(1, parts, lambda df: df.axes[1])
 
-    index = partition_mgr_class.get_indices(0, parts, lambda df: df.axes[0])
-    columns = partition_mgr_class.get_indices(1, parts, lambda df: df.axes[1])
     return DataFrame(
-        query_compiler=PandasQueryCompiler(partition_frame_class(parts, index, columns))
+        query_compiler=PandasQueryCompiler(
+            partition_frame_class(
+                parts,
+                index,
+                columns,
+                row_lengths=row_lengths,
+                column_widths=column_widths,
```

Comment:
Let's also add `dtypes` parameter.

Reply:
Though, we can wait until it is really required. Please, fix small fix about the supplementation and we will merge the PR.

Reply:
It's already added. The change doesn't look costly. Maybe let's leave him?

Reply:
We can leave it but in that case we need to add a note on `dtypes` must match underlying data of the partitions, i.e. having passed `dtypes` the DataFrame columns won't be cast to that types. Just to avoid confusing in such cases:
```python
import pandas
df = pandas.DataFrame([1, 2, 3], dtype="float")
df
     0
0  1.0
1  2.0
2  3.0
```
 Also, passing `dtypes` doesn't trigger computations when creating a DataFrame.

Reply:
It that case let's postpone this until it's really required, as you proposed.

Reply:
Ok

Reply:
Changes related to `dtypes` were removed.